### PR TITLE
feat: Solve Keccak opcodes

### PR DESCRIPTION
**Pull request by guipublic**
# Related issue(s)

Related to (https://github.com/noir-lang/noir/issues/1189)

# Description

## Summary of changes

Solve Keccak opcodes

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

### CODE REVIEW

File: acvm/src/pwg/hash.rs [hunk 1]
```diff
@@ -72,3 +73,42 @@ fn generic_hash_256<D: Digest>(
     let result = hasher.finalize().as_slice().try_into().unwrap();
     Ok(result)
 }
+
+pub fn keccak256(
+    initial_witness: &mut BTreeMap<Witness, FieldElement>,
+    gadget_call: &BlackBoxFuncCall,
+) -> Result<(), OpcodeResolutionError> {
```
Comment (TomAFrench) [inline acvm/src/pwg/hash.rs:80]:
```suggestion
) -> Result<OpcodeResolution, OpcodeResolutionError> {
```

This needs to match the interface for the other bb function solvers

File: acvm/src/pwg/hash.rs [hunk 2]
```diff
@@ -72,3 +73,42 @@ fn generic_hash_256<D: Digest>(
...
+    generic_sha3::<Keccak256>(initial_witness, gadget_call)
+}
+
+fn generic_sha3<D: sha3::Digest>(
```
Comment (TomAFrench) [inline acvm/src/pwg/hash.rs:84]: This is the same `Digest` trait as used by `generic_hash_256` so we don't need a duplicate implementation ( you may need to adjust versions of dependencies)

File: acvm/src/lib.rs [hunk 3]
```diff
@@ -122,7 +124,12 @@ pub trait PartialWitnessGenerator {
                                 unassigned_witness.0,
                             )))
                         } else {
-                            self.solve_black_box_function_call(initial_witness, bb_func)
+                            let status = solve_black_box_function(initial_witness, bb_func);
+                            if matches!(status, Err(OpcodeResolutionError::OpcodeNotSolvable(_))) {
+                                self.solve_black_box_function_call(initial_witness, bb_func)
+                            } else {
+                                status
+                            }
```
Comment (TomAFrench) [inline acvm/src/lib.rs:132]: This change is unrelated to supporting keccak opcodes and has significant knock on effects for all backends so shouldn't be included in this PR.

**Review by TomAFrench — CHANGES_REQUESTED** [review]

File: acvm/src/lib.rs (hunk 3 above)
Comment (guipublic) [reply acvm/src/lib.rs:132]: I don't think it is a breaking change, what is broken?

File: acvm/src/pwg/hash.rs (hunk 2 above)
Comment (guipublic) [reply acvm/src/pwg/hash.rs:84]: It is not the same because the two crates do not use the same version.

File: acvm/src/pwg/hash.rs (hunk 1 above)
Comment (guipublic) [reply acvm/src/pwg/hash.rs]: Done

File: acvm/src/pwg/hash.rs (hunk 2 above)
Comment (TomAFrench) [reply acvm/src/pwg/hash.rs:85]: Indeed, you'll need to adjust the dependency version for the blake2 crate to match.

File: acvm/src/lib.rs (hunk 3 above)
Comment (TomAFrench) [reply acvm/src/lib.rs:132]: It's changing the contract between backends and ACVM.

Previously ACVM provided reference methods for executing black box functions to the backends however backends were in charge of solving the black box function and returning the result. After this change this is still the case for all other opcodes but we'll never send keccak256 in particular to the backend.

This split between keccak256 and all the other functions doesn't make sense and we shouldn't split responsibility for black box function execution between the backend and the ACVM. Until we have a rust implementation of all black box functions then we should allow the backend to handle this.

Comment (guipublic) [reply acvm/src/lib.rs:132]: So in fact this is not breaking anything.
Solving Keccak in ACVM does make sense, as well as all other 'blackbox'. I will move the solving into ACVM for all but pedersen, which will still be handled by the backend until we have a rust implementation.

Comment (phated) [reply acvm/src/lib.rs:132]: This change would mean that a lot of code in aztec_backend would become dead code without us actually realizing it. The backend has a match on all black box functions but those code paths would never be executed.

This could lead to confusing bugs between Noir and the backend and it seems hard to communicate to backend implementors.

I'm torn on where this should happen because I don't want aztec_backend to implement keccak itself 😬 I'd probably be fine making a separate PR for this change, just so we can mark it breaking and have a clear CHANGELOG entry. @TomAFrench thoughts?

Comment (guipublic) [reply acvm/src/lib.rs:132]: Of course we would remove the dead code as long as it is integrated in ACVM, so I don't see this as an issue.

Comment (phated) [reply acvm/src/lib.rs:132]: > Of course we would remove the dead code as long as it is integrated in ACVM, so I don't see this as an issue.

This assumes that we are in control of all the backends, but we need to develop as though there are other backends that need to integrate ACVM and they will be confused without accurate documentation & changelog entries.

**Comment by kevaundray** [comment]
I'm not a fan of PRs that don't say what they do in the title, since it makes it harder to find where changes happened and why and it also invites feature creap into PRs.

Lets just add Keccak into this PR exactly like how it is done for other functions, then any other changes can be done in a separate PR and discussed.

File: acvm/src/lib.rs (hunk 3 above)
Comment (phated) [reply acvm/src/lib.rs:132]: We could land something like https://github.com/noir-lang/acvm/pull/257 before this and then it wouldn't be an issue.

**Comment by TomAFrench** [comment]
Closing this as it's been superceded by #259 and #264
